### PR TITLE
LibWeb: Introduce RecordingPainter to serialize painting commands

### PR DESCRIPTION
**Pull request by kalenikaliaksandr**
This modification introduces a new layer to the painting process. The stacking context traversal no longer immediately calls the Gfx::Painter methods. Instead, it writes serialized painting commands into newly introduced RecordingPainter. Created list of commands is executed later to produce resulting bitmap.

Producing painting command list will make it easier to add new optimizations:
- It's simpler to check if the painting result is not visible in the viewport at the command level rather than during stacking context traversal.
- Run painting in a separate thread. The painting thread can process serialized painting commands, while the main thread can work on the next paintable tree and safely invalidate the previous one.
- As we consider GPU-accelerated painting support, it would be easier to back each painting command rather than constructing an alternative for the entire Gfx::Painter API.

### CODE REVIEW

File: Userland/Services/WebContent/PageHost.cpp [hunk 1]
```diff
@@ -121,7 +121,8 @@ Gfx::Color PageHost::background_color() const
 
 void PageHost::paint(Web::DevicePixelRect const& content_rect, Gfx::Bitmap& target)
 {
-    Gfx::Painter painter(target);
+    auto bitmap = Gfx::Bitmap::create(Gfx::BitmapFormat::BGRA8888, content_rect.size().to_type<int>()).release_value_but_fixme_should_propagate_errors();
+    Gfx::Painter painter(bitmap);
```
Comment (ADKaster) [inline Userland/Services/WebContent/PageHost.cpp:125]: This looks unused? 

The most drive-by review ever btw. I didn't look at any file but this one :D

Comment (kalenikaliaksandr) [reply Userland/Services/WebContent/PageHost.cpp]: indeed! thank you, I removed it 😅

File: Userland/Libraries/LibWeb/Painting/RecordingPainter.cpp [hunk 2]
```diff
@@ -0,0 +1,829 @@
...
+                auto sy = static_cast<float>(source_rect.height()) / transformed_destination_rect.height();
+                bitmap = TRY(bitmap->scaled(sx, sy));
+                destination_clipped_fixup.scale_by(sx, sy);
+            }
+            return bitmap;
+        };
+
+        auto bitmap_or_error = try_get_scaled_destination_bitmap();
+        if (bitmap_or_error.is_error()) {
+            // NOTE: If creation of bitmap failed we need to skip all painting commands that belong to this stacking context.
```
Comment (awesomekling) [inline Userland/Libraries/LibWeb/Painting/RecordingPainter.cpp:208]: I think it would be okay if we require that allocating the bitmap succeeds here.
If it fails, we will not render the page correctly, and there is no in-spec mechanism for recovering and/or reporting this kind of problem.

If we make this temporary bitmap allocation infallible, it simplifies the display list iteration and makes it less confusing how traversal works :)

Comment (kalenikaliaksandr) [reply Userland/Libraries/LibWeb/Painting/RecordingPainter.cpp]: Actually `get_region_bitmap()` has a bit confusing api and returns an error if requested bitmap is located outside of viewport (has 0x0 size) so we should proceed executing painting commands in that case. I added a FIXME that explains that.

For now I changed all commands to return `Continue` or `SkipStackingContext` and let main commands execution loop handle skipping. I hope it makes `PushStackingContext` less confusing :)
